mayastor/src/bdev/nexus/nexus_metadata.rs

```diff
@@ -0,0 +1,620 @@
+            buf.as_slice(),
+            header.used_entries,
+        )?;
+        let checksum = MetaDataIndexEntry::checksum(&index);
+        if checksum != header.index_checksum {
+            return Err(MetaDataError::IndexChecksum {});
+        }
+        Ok(index)
+    }
+
+    /// Construct and validate an empty index
+    pub fn empty_index(
+        header: &MetaDataHeader,
+    ) -> Result<Vec<MetaDataIndexEntry>, MetaDataError> {
+        if header.index_checksum != 0 {
+            return Err(MetaDataError::IndexChecksum {});
+        }
+        Ok(Vec::new())
+    }
+
+    /// Construct a (config) object from raw data
+    pub fn read_config_object(
+        buf: &DmaBuf,
+        entry: &MetaDataIndexEntry,
+    ) -> Result<NexusConfig, MetaDataError> {
+        let checksum = crc32::checksum_ieee(buf.as_slice());
+        if checksum != entry.data_checksum {
+            return Err(MetaDataError::ObjectChecksum {});
+        }
+        Ok(NexusConfig::from_slice(buf.as_slice())
+            .context(DeserializeError {})?)
+    }
+}
+
+impl NexusChild {
+    async fn write_index(
+        &self,
+        metadata: &NexusMetaData,
+    ) -> Result<(), MetaDataError> {
+        let (bdev, _desc) = self.get_dev()?;
+        let block_size = bdev.block_len() as u64;
+
+        let blocks = metadata.header.index_start
+            + Unsigned::get_aligned_blocks(
+                (metadata.header.max_entries * metadata.header.entry_size)
+                    as u64,
+                block_size,
+            );
+        let mut buf =
+            DmaBuf::new((blocks * block_size) as usize, bdev.alignment())
+                .context(IndexAlloc {})?;
+        let mut writer = Cursor::new(buf.as_mut_slice());
+
+        // Header
+        serialize_into(&mut writer, &metadata.header)
+            .context(SerializeError {})?;
+
+        // Index
+        writer
+            .seek(SeekFrom::Start(metadata.header.index_start * block_size))
+            .unwrap();
+        for entry in &metadata.index {
+            serialize_into(&mut writer, entry).context(SerializeError {})?;
+        }
+
+        self.write_at(metadata.header.self_lba * block_size, &buf)
+            .await
+            .context(IndexWrite {})?;
+
+        Ok(())
+    }
+
+    pub async fn write_config_object(
+        &self,
+        metadata: &mut NexusMetaData,
+        config: &NexusConfig,
+        now: &SystemTime,
+    ) -> Result<(), MetaDataError> {
+        if metadata.header.used_entries >= metadata.header.max_entries {
+            return Err(MetaDataError::IndexSizeExceeded {
+                count: metadata.header.used_entries + 1,
+                size: metadata.header.max_entries,
+            });
+        }
+
+        let (bdev, _desc) = self.get_dev()?;
+        let block_size = bdev.block_len() as u64;
+
+        let timestamp = now
+            .duration_since(UNIX_EPOCH)
+            .context(TimeStampError {})?
+            .as_secs();
```

Comment:
I think it would be a good idea to have a higher resolution here (i.e usec or nsec?)